transport/http2_client.go

```diff
@@ -1068,37 +1091,112 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
-			case <-t.writableChan:
-				switch i := i.(type) {
-				case *windowUpdate:
-					t.framer.writeWindowUpdate(true, i.streamID, i.increment)
-				case *settings:
-					if i.ack {
-						t.framer.writeSettingsAck(true)
-						t.applySettings(i.ss)
-					} else {
-						t.framer.writeSettings(true, i.ss...)
-					}
-				case *resetStream:
-					t.framer.writeRSTStream(true, i.streamID, i.code)
-				case *flushIO:
-					t.framer.flushWrite()
-				case *ping:
-					t.framer.writePing(true, i.ack, i.data)
-				default:
-					grpclog.Printf("transport: http2Client.controller got unexpected item type %v\n", i)
+			// If controlMsg is of type resetKeepaliveTimer,
+			// then check if the keepaliveSkipped flag is still set.
+			if _, ok := controlMsg.(resetKeepaliveTimer); ok {
+				atomic.StoreUint32(&t.keepaliveSkipped, 0)
+				// Reset the timer to timeout.
+				// Note : This is safe to read, since the
+				// only codepath that sets the keepaliveSkipped
+				// flag also resets the timer to infinity.
+				// Thus, there'll never be a case where we are
+				// trying to read from an empty timer channel.
+				isPingSent = true
+				if !timer.Stop() {
+					<-timer.C
 				}
-				t.writableChan <- 0
+				timer.Reset(t.kp.Timeout)
+				continue
+			}
+			wchan = t.writableChan
+			cchan = nil
+		case <-wchan:
+			switch i := controlMsg.(type) {
+			case *windowUpdate:
+				t.framer.writeWindowUpdate(true, i.streamID, i.increment)
+			case *settings:
+				if i.ack {
+					t.framer.writeSettingsAck(true)
+					t.applySettings(i.ss)
+				} else {
+					t.framer.writeSettings(true, i.ss...)
+				}
+			case *resetStream:
+				t.framer.writeRSTStream(true, i.streamID, i.code)
+			case *flushIO:
+				t.framer.flushWrite()
+			case *ping:
+				t.framer.writePing(true, i.ack, i.data)
+			default:
+				grpclog.Printf("transport: http2Client.controller got unexpected item type %v\n", i)
+			}
+			wchan <- 0
+			wchan = nil
+			cchan = t.controlBuf.get()
+		case <-timer.C:
+			// All code paths in this case must reset the timer.
+
+			// Get the activity counter value and reset it.
+			a := atomic.SwapUint64(&t.activity, 0)
+			if a > 0 {
+				atomic.StoreUint32(&t.keepaliveSkipped, 0)
+				isPingSent = false
+				timer.Reset(t.kp.Time)
 				continue
-			case <-t.shutdownChan:
-				return
 			}
+			if isPingSent {
+				t.Close()
+				timer.Reset(infinity)
+				continue
+			}
+			t.mu.Lock()
+			ns := len(t.activeStreams)
+			if !t.kp.PermitWithoutStream && ns < 1 {
+				// set flag that signifyies keepalive was skipped
+				atomic.StoreUint32(&t.keepaliveSkipped, 1)
+				t.mu.Unlock()
+				timer.Reset(infinity)
+				continue
+			}
+			t.mu.Unlock()
+			// reset the keepaliveSkipped flag
+			atomic.StoreUint32(&t.keepaliveSkipped, 0)
+			// Send ping.
+			t.controlBuf.put(keepalivePing)
+			isPingSent = true
+			timer.Reset(t.kp.Timeout)
 		case <-t.shutdownChan:
+			// stop the keepalive timer
+			if !timer.Stop() {
+				select {
+				case <-timer.C:
+				default:
```

Comment:
Since we exit the controller function right after, wouldn't the timer be eventually gc'd. Also, since nowhere outside the function is timer.C accessed, it should be safe even if timer.C later gets a value. In fact, maybe we can get rid of the case <- timer.c: altogether?